### PR TITLE
CORE-20100: Retry logic around finality and external calls from a flow

### PR DESCRIPTION
**Pull request by JamesHR3**
### Problem description

Failures in `FinalityFlow` owing to transient issues such as database connection errors can cause the ledger to become out of sync. While this can be repaired after the fact, the platform should aim to prevent this from happening as far as is possible.

### Solution

Introduce replay of events on failure within the event mediator, categorize errors to determine how frequently to replay, and bound replays of anything encountering unknown errors.

The aim is to provide a balance between pushing as many flows in finality to successful completion as possible, while not allowing broken flows to block the system for too long.

The actual fix consists of the following:
- Logic in the event mediator to retry events if errors are reported back from event processing. Errors may come back as either transient or fatal.
- Reworked logic in the `RPCClient` to mark particular errors as transient or fatal.
- Reworked categorization logic in the persistence processor to ensure persistence errors are reported correctly. (This addresses at least one bug where an error to be returned to the user instead killed the flow.)
- Infinite retry logic within `FinalityFlow` itself. This should prevent session errors from the other side terminating finality early when it is inappropriate to do so.

### CODE REVIEW

File: components/persistence/persistence-service-common/src/main/kotlin/net/corda/persistence/common/PersistenceExceptionCategorizerImpl.kt [hunk 1]
```diff
@@ -18,24 +22,82 @@ import javax.persistence.RollbackException
...
+            criteria<SQLException> {
+                it.sqlState in setOf("08001", "08003", "08004", "08006", "08006", "58030")
+            },
+            criteria<SQLException> {
+                it.message == "Connection is closed"
+            },
+            criteria<SocketException>()
+        )
+        return checks.any { it.meetsCriteria(exception) }
+/*        return when (exception) {
```
Comment (LWogan) [inline components/persistence/persistence-service-common/src/main/kotlin/net/corda/persistence/common/PersistenceExceptionCategorizerImpl.kt]: Some commented out code here

File: libs/membership/membership-common/src/main/kotlin/net/corda/membership/lib/MemberInfoExtension.kt [hunk 2]
```diff
@@ -301,10 +302,10 @@ class MemberInfoExtension {
             key: String,
             keyListGetter: () -> Collection<PublicKey>
         ) = memberProvidedContext.parseSet<SecureHash>(key).ifEmpty {
-            logger.warn(
+            logger.debug {
```
Comment (LWogan) [inline libs/membership/membership-common/src/main/kotlin/net/corda/membership/lib/MemberInfoExtension.kt:305]: I noticed this was quite spammy in the test runs.  Was that due to how our tests run or is it something that would be common in production. Is this change ok to make debug.

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/ConsumerProcessor.kt [hunk 3]
```diff
@@ -54,6 +53,10 @@ class ConsumerProcessor<K : Any, S : Any, E : Any>(
 
     private val stateManager = config.stateManager
 
+    companion object {
+        private const val MAX_FAILURE_ATTEMPTS = 5
```
Comment (LWogan) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/ConsumerProcessor.kt:57]: We have a tendency in the mediator to hardcode things we later wish to be configurable. Are we sure we want to hardcode this?

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/ConsumerProcessor.kt [hunk 4]
```diff
@@ -93,14 +97,122 @@ class ConsumerProcessor<K : Any, S : Any, E : Any>(
...
+
     /**
      * Polls records from the message bus, assigns records into groups, retrieves states for the records and then passses each group to
      * a thread for processing by the [eventProcessor].
      * Sends any async outputs back to the bus and saves states to the state manager.
      * Retries processing record keys whose states failed to save.
      * @param consumer message bus consumer
      */
-    private fun pollAndProcessEvents(consumer: MediatorConsumer<K, E>) {
+/*    private fun pollAndProcessEvents(consumer: MediatorConsumer<K, E>) {
```
Comment (LWogan) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/ConsumerProcessor.kt]: commented out code

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/ConsumerProcessor.kt [hunk 5]
```diff
@@ -93,14 +97,122 @@ class ConsumerProcessor<K : Any, S : Any, E : Any>(
         consumer?.close()
     }
 
+    private fun pollLoop(consumer: MediatorConsumer<K, E>, failureCounts: MutableMap<String, Int>) {
```
Comment (LWogan) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/ConsumerProcessor.kt:117]: Need some docs on these methods

**Review by LWogan — COMMENTED** [review]
Early review of the draft PR.

The code changes looks pretty good to me. I don't see anything major wrong here. However I'm not familiar enough with the error types to comment on the categorization.

File: libs/membership/membership-common/src/main/kotlin/net/corda/membership/lib/MemberInfoExtension.kt (hunk 2 above)
Comment (JamesHR3) [reply libs/membership/membership-common/src/main/kotlin/net/corda/membership/lib/MemberInfoExtension.kt:305]: I suspect this is very common. `WARN` is inappropriate for this, as it does not represent something that will cause a failure of processing.

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/ConsumerProcessor.kt (hunk 3 above)
Comment (JamesHR3) [reply libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/ConsumerProcessor.kt:57]: I'd go the other way. It's much harder to remove config we don't need than it is to add config we do (due to backwards compatibility concerns). We already have plenty of examples of "dead" config, particularly in the messaging layer.

In the event we need this to be configurable later, we can make it so. I don't believe we have a good use case for doing that yet.

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/ConsumerProcessor.kt (hunk 5 above)
Comment (JamesHR3) [reply libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/ConsumerProcessor.kt:117]: I've added some comments above these functions, let me know what you think.

**Review by LWogan — APPROVED** [review]
LGTM

**Review by rick-r3 — APPROVED** [review]
LGTM

**Pull request merged**